Add empty design (#1)
whitepau/oneAPI-samples: DirectProgramming/DPC++FPGA/Tutorials/GettingStarted/fpga_template/README.md

# FPGA Template

This project serves as a template for Intel® oneAPI FPGA designs. 

| Optimized for                     | Description
|:---                               |:---
| OS                                | Linux* Ubuntu* 18.04/20.04 <br> RHEL*/CentOS* 8 <br> SUSE* 15 <br> Windows* 10

> | Hardware                          | Intel® FPGA Programmable Acceleration Card (PAC) D5005 (with Intel Stratix® 10 SX) <br> Intel® FPGA 3rd party / custom platforms with oneAPI support <br> *__Note__: Intel® FPGA PAC hardware is only compatible with Ubuntu 18.04*

Author: I intentionally removed 10 PAC since this design depends on USM, so it won't work with A10 PAC.

> | Hardware                          | Intel® FPGA Programmable Acceleration Card (PAC) D5005 (with Intel Stratix® 10 SX) <br> Intel® FPGA 3rd party / custom platforms with oneAPI support <br> *__Note__: Intel® FPGA PAC hardware is only compatible with Ubuntu 18.04*

Author: I intentionally omitted A10 PAC because this design depends on USM so A10 won't work.

> | Hardware                          | Intel® FPGA Programmable Acceleration Card (PAC) D5005 (with Intel Stratix® 10 SX) <br> Intel® FPGA 3rd party / custom platforms with oneAPI support <br> *__Note__: Intel® FPGA PAC hardware is only compatible with Ubuntu 18.04*

Author: as discussed, we will use USM since it's better for IPA

| Software                          | Intel® oneAPI DPC++ Compiler <br> Intel® Quartus Prime Pro Edition <br> Siemens® Questa® Intel® FPGA Starter Edition
| What you will learn               | Best practices for creating and managing a oneAPI FPGA project
| Time to complete                  | 10 minutes

## Purpose

> Use this project as a starting point when you build designs for the Intel® oneAPI FPGA compiler. It includes a CMake build system to abstract away the need to know all the various command-line flags for the `icpx` compiler, and a simple single-source design to serve as an example. You can customize the build flags by modifying the top part of `src/CMakeLists.txt`. If you want to pass additional flags to Intel DPC++ compiler, you can change the `USER_FLAGS` and `USER_HARDWARE_FLAGS` variables defined in `src/CMakeLists.txt`. 

Author: good catch. I blame our marketing folks for being confusing :)

|-----------------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| `USER_HARDWARE_FLAGS` | These flags apply only to flows that generate FPGA hardware (i.e. report, simulation, hardware). You can specify flags such as `-Xsclock` or `-Xshyper-optimized-handshaking=off` |
| `USER_FLAGS`          | These flags apply to all flows, including emulation. You can specify flags such as `-v` or define macros such as `-DYOUR_OWN_MACRO=3`                                             |

```bash
###############################################################################
### Customize these build variables
###############################################################################
set(SOURCE_FILE vector_add.cpp)
set(TARGET_NAME vector_add)

# When targeting a board, the `FPGA_BOARD` will be targeted. Options are:
#   intel_s10sx_pac:pac_s10
#   intel_s10sx_pac:pac_s10_usm
#   intel_a10gx_pac:pac_a10

> set(FPGA_BOARD "intel_s10sx_pac:pac_s10_usm")

Author: OK i think I like this. My intention had been to set a 'board target' and a 'device target' and then let the user choose to build for a board or build for an IP, but i suppose we can restrict the build targets and select 'board' or 'device' based off whatever the target the user chooses is. makefile targets would be:
- `fpga_emu`
- `report`
- `fpga_sim`
- `fpga_ip_export` (add once we support IP authoring; this will effectively just be the same as `report`, but with `-fsycl-device-code-split=per_kernel`).
- `fpga` (will run quartus for both IP Authoring and full-system)

set(USER_HARDWARE_FLAGS "")
# use cmake -DUSER_HARDWARE_FLAGS=<flags> to set extra flags for FPGA backend
# compilation

set(USER_FLAGS "")
# use cmake -DUSER_FLAGS=<flags> to set extra flags for general compilation
```

Everything below this in the `src/CMakeLists.txt` is necessary for selecting the compiler flags that are necessary to support the build targets specified below, and should not need to be modified.

> ## License

Author: I do not understand; this paragraph indicates to the user how they should modify the CMakeLists.txt; are you recommending that I move the entire listing from the CMakeLists.txt to the end? that belongs in the 'purpose' section with the description.

> ## License

Reviewer: I meant the "License" section :)

Code samples are licensed under the MIT license. See
[License.txt](https://github.com/oneapi-src/oneAPI-samples/blob/master/License.txt) for details.

Third party program Licenses can be found here: [third-party-programs.txt](https://github.com/oneapi-src/oneAPI-samples/blob/master/third-party-programs.txt).

## Building the FPGA Template Tutorial
Use these commands to run the design, depending on your OS.

### On a Linux* System 
This design uses CMake to build the project, generate build artifacts for GNU/make like this:

> 

Author: applied this, except added a note that Arria10 PAC is not supported due to USM.

```bash
mkdir build
cd build
cmake ..
```

> 

Reviewer: ```suggestion

2. Compile the design through the generated `Makefile`. The following build targets are provided, matching the recommended development flow:
```

| Target          | Expected Time  | Output                                                        | Description                                                                                                                                                                                                                                                                                             |
|-----------------|----------------|---------------------------------------------------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| `make fpga_emu` | Seconds        | x86-64 binary                                                 | Compiles the FPGA device code to the CPU. Use the Intel® FPGA Emulation Platform for OpenCL™ software to verify your SYCL code’s functional correctness.                                                                                                                                                |

> | `make report`   | Minutes        | RTL                                                           | Compiles the FPGA device code to RTL and generates an optimization report that describes the structures generated on the FPGA, identifies performance bottlenecks, and estimates resource utilization. This report will include the interfaces defined in your selected Board Support Package.          |

Author: added reports for the appropriate ones as well. Actually, to open the reports now they are recommending user to use `fpga_report` (`fpga_report.exe` on Windows) to view the report; it appears in `build/<a>.prj/reports/<a>_report.zip`. I've added a note about this as well.

| `make fpga`     | Multiple Hours | Quartus Place & Route (Full accelerator) + x86-64 host binary | Compiles the FPGA device code to RTL and generate an FPGA image that you can run on a supported accelerator board.                                                                                                                                                                                      |
| `make fpga_sim` | Minutes        | RTL + x86-64 binary                                           | Compiles the FPGA device code to RTL and generates a simulation testbench. Use the Questa*-Intel® FPGA Edition simulator to verify your design.                                                                                                                                                         |

The `fpga_emu`, `fpga_sim` and `fpga` targets produce binaries that you can run. The executables will be called `TARGET_NAME.fpga_emu`, `TARGET_NAME.fpga_sim`, and `TARGET_NAME.fpga`, where `TARGET_NAME` is the value you specify in `src/CMakeLists.txt`.

### On a Windows* System

> This design uses CMake to build the project, and generate build artifacts for `nmake` like this:

Reviewer: ```suggestion
This design uses CMake to build the project, and generate build artifacts for `nmake`.

1. Generate the `Makefile` by running `cmake`.
```

> 

Reviewer: The following bash commands should be replaced with:
   ```
   mkdir build
   cd build
   ```
   To compile for the Intel® PAC with Intel Arria® 10 GX FPGA, run `cmake` using the command:
   ```
    cmake -G "NMake Makefiles" ..
   ```
   Alternatively, to compile for the Intel® FPGA PAC D5005 (with Intel Stratix® 10 SX), run `cmake` using the command:

   ```
   cmake -G "NMake Makefiles" .. -DFPGA_BOARD=intel_s10sx_pac:pac_s10
   ```
   You can also compile for a custom FPGA platform. Ensure that the board support package is installed on your system. Then run `cmake` using the command:
   ```
   cmake -G "NMake Makefiles" .. -DFPGA_BOARD=<board-support-package>:<board-variant>
   ```

```bash
mkdir build
cd build
cmake -G "NMake Makefiles" ..
```

> 

Reviewer: ```suggestion

2. Compile the design through the generated `Makefile`. The following build targets are provided, matching the recommended development flow:
```

> This project can build 8 targets for Windows.

Reviewer: ```suggestion
This project can build 4 targets for Windows.
```

| Target           | Expected Time  | Output                                                        | Description                                                                                                                                                                                                                                                                                    |
|------------------|----------------|---------------------------------------------------------------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| `nmake fpga_emu` | Seconds        | x86-64 binary                                                 | Compiles the FPGA device code to the CPU. Use the Intel® FPGA Emulation Platform for OpenCL™ software to verify your SYCL code’s functional correctness.                                                                                                                                       |

> | `nmake report`   | Minutes        | RTL                                                           | Compiles the FPGA device code to RTL and generates an optimization report that describes the structures generated on the FPGA, identifies performance bottlenecks, and estimates resource utilization. This report will include the interfaces defined in your selected Board Support Package. |

Author: noted above

| `nmake fpga`     | Multiple Hours | Quartus Place & Route (Full accelerator) + x86-64 host binary | Compiles the FPGA device code to RTL and generate an FPGA image that you can run on a supported accelerator board.                                                                                                                                                                             |

> | `nmake fpga_sim` | Minutes        | RTL + x86-64 binary                                           | Compiles the FPGA device code to RTL and generates a simulation testbench. Use the Questa*-Intel® FPGA Edition simulator to verify your design.                                                                                                                                                |

Author: now it's called questa

> 

Reviewer: ```suggestion

> **Note**: The Intel® PAC with Intel Arria® 10 GX FPGA and Intel® FPGA PAC D5005 (with Intel Stratix® 10 SX) do not support Windows*. Compiling to FPGA hardware on Windows* requires a third-party or custom Board Support Package (BSP) with Windows* support.

> **Note**: If you encounter any issues with long paths when compiling under Windows*, you may have to create your ‘build’ directory in a shorter path, for example c:\samples\build.  You can then run cmake from that directory, and provide cmake with the full path to your sample directory.

```

> The `fpga_emu`, `fpga_sim`, and `fpga` targets also produce binaries that you can run. The executables will be called `TARGET_NAME.fpga_emu.exe`, `TARGET_NAME.fpga_sim.exe`, and `TARGET_NAME.fpga.exe`, where `TARGET_NAME` is the value you specify in `src/CMakeLists.txt`.

Author: I thought you had removed the need for these in your PR we talked about on Wednesday?

> The `fpga_emu`, `fpga_sim`, and `fpga` targets also produce binaries that you can run. The executables will be called `TARGET_NAME.fpga_emu.exe`, `TARGET_NAME.fpga_sim.exe`, and `TARGET_NAME.fpga.exe`, where `TARGET_NAME` is the value you specify in `src/CMakeLists.txt`.

Reviewer: I got the approval to do so - the new top level readme will be merged this week, and we'll clean up all code samples individual README in a bulk PR after wards to make them all consistent.
So for now it makes sense to have them, so that they are bulk removed/replaced with a pointer to the top level readme afterwards
